### ISSUE TITLE
'formatoptions' is not respected.

### ISSUE THREAD

**Issue by kevincox** [autozimu/LanguageClient-neovim#821]
## Describe the bug

When [using LanguageClient-neovim as `formatexpr`](https://github.com/autozimu/LanguageClient-neovim/blob/53ce2fee7d24dd9262ea783908e2edebc71635dc/doc/LanguageClient.txt#L64) `formatoptions` configuration is not respected.

One example of a really annoying case is if `textwidth` is set vim will call `formatexpr` once the line is over the "limit" however with `formatoptions-=t` wrapping should not be performed in non-comment areas. However with these two options set formatting is still performed (additionally the cursor is left in the wrong place leading to a big mess).

## Environment
- neovim/vim version (`nvim --version` or `vim --version`):

```
% nvim --version
NVIM v0.3.4
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wconversion -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.3.4/src -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Features: +acl +iconv +jemalloc +tui 
```

Also tested with vim 8.1.1073.

- This plugin version (`git rev-parse --short HEAD`): 5685989
- This plugin's binary version (`bin/languageclient --version`): `languageclient 0.1.140`
- Minimal vimrc content (A minimal vimrc is the smallest vimrc that could
  reproduce the issue. Refer to an example [here][min-vimrc.vim]):

```
set textwidth=80
set formatoptions=
set formatexpr=LanguageClient#textDocument_rangeFormatting_sync()
let g:LanguageClient_serverCommands = { 'rust': ['rustup', 'run', 'stable', 'rls'] }
```

## Current behavior

Example undesirable behaviour, `|` represents cursor.

```rust
# Before entering `)`
assert_eq!(trace, &[crate::Position::new(4, 1), crate::Position::new(4, 8),]|
# After entering `)`
assert_eq!()|
	trace,
	&[crate::Position::new(4, 1), crate::Position::new(4, 8),]
```

## Expected behavior

No formatting occurs when not matching `formatoptions`.

## Additional context

This seems like a really weird design choice on the VIM side to even call the `formatexpr` at all. However I guess I can see some benefits such as the `formatexpr` being able to decide what is a comment and what is "text". Maybe this could be pushed "upstream" to be fixed in vim.

**Comment by autozimu** [comment]
Sorry for the inconvenience.

While in the case, the root cause is that language server/LSP don't have any specialized knowledge about vim's `textwidth` nor `formatoptions`, thus the formatted result won't respect those vim settings at all.

The specification does leave room for additional property for formatting,
```typescript
	/**
	 * Signature for further properties.
	 */
	[key: string]: boolean | number | string;
``` 
But again language server should made to understand additional properties like `textwidth`.

**Comment by kevincox** [comment]
I agree that for the complex cases we would want to add this to the protocol. However for the simple cases it probably makes more sense to handle it in the client.

- If formatting is disabled it should not contact the server.
- If we are in a comment and comment formatting is disabled it should not contact the server.
- If we don't support keeping the cursor position we should probably never format.

I think an ok short term workaround to skip formatting if `v:char` is set. This way no auto-formatting will be performed. I strongly feel that not breaking people's code is more important than auto-formatting while typing.
